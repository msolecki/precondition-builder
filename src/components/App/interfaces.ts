export interface LocationInterface {
    radius: number | null;
    latLng: string | null;
}

enum OperatingSystemInterface {
    iOS = 'iOS',
    Android = 'Android',
}

export interface NuggetConditionInterface {
    id: number;
    condition: 'read';
    value: boolean;
}

export interface ConditionInterface {
    logged: boolean | null;
    location: LocationInterface | null;
    nuggets: NuggetConditionInterface[] | null;
}
