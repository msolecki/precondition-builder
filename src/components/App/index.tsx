import React, {ChangeEvent, ReactElement} from 'react'
import {Container} from '@material-ui/core'
import AppBar from '@material-ui/core/AppBar'
import Tabs from '@material-ui/core/Tabs'
import Tab from '@material-ui/core/Tab'
import Typography from '@material-ui/core/Typography'
import AdvancedTab from '../Tabs/AdvancedTab'
import BasicTab from '../Tabs/BasicTab'
import {ConditionInterface} from './interfaces'

export interface StateInterface {
    conditions: ConditionInterface[];
    tabNumber: number;
}

class App extends React.Component<{}, StateInterface> {
    public constructor(props: {}) {
        super(props)

        this.state = {
            conditions: [],
            tabNumber: 0
        }
    }

    private changeTab = (event: ChangeEvent<{}>, newNumber: number): void => {
        this.setState({tabNumber: newNumber})
    }

    public handleAddCondition = (newCondition: ConditionInterface): void => {
        const conditions = this.state.conditions
        conditions.push(newCondition)

        this.setState({conditions})
    }

    public render(): ReactElement {
        const {tabNumber, conditions} = this.state

        return (
            <Container style={{width: '50%'}}>
                <AppBar position="static">
                    <Tabs value={tabNumber} onChange={this.changeTab} variant="fullWidth">
                        <Tab label="Basic"/>
                        <Tab label="Advanced"/>
                    </Tabs>
                </AppBar>
                <Typography component="div" style={{paddingTop: 25}}>
                    {tabNumber === 0 && <BasicTab handleAddCondition={this.handleAddCondition} conditions={conditions}/>}
                    {tabNumber === 1 && <AdvancedTab data={conditions}/>}
                </Typography>
            </Container>
        )
    }
}

export default App
