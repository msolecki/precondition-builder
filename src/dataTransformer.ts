import {ConditionInterface, NuggetConditionInterface} from './components/App/interfaces'

export const transform = (conditions: ConditionInterface[], activated: boolean): object => {
    return {
        active: activated,
        conditions: conditions.map((condition: ConditionInterface): string[] => {
            const result: string[] = []

            if (condition.system) {
                result.push(`client.tracking.system.os == '${condition.system}'`)
            }

            if (condition.logged !== null) {
                result.push(`logged == '${condition.logged}'`)
            }

            if (condition.location && condition.location.radius) {
                result.push(`location_radius == '${condition.location.radius}'`)
            }

            if (condition.nuggets) {
                condition.nuggets.forEach((nugget: NuggetConditionInterface): void => {
                    result.push(`public.kesi.additionalData.${nugget.id}.${nugget.condition} == ${nugget.value}`)
                })
            }

            return result
        }).filter((item): boolean => item.length > 0)
    }
}
