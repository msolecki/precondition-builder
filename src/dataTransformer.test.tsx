import * as React from 'react'
import {ConditionInterface} from './components/App/interfaces'
import {transform} from './dataTransformer'

const createConditions = (logged, system, location, nuggets): ConditionInterface => {
    return {
        logged,
        system,
        location,
        nuggets
    }
}

describe('dataTransformer', () => {
    describe('empty conditions', () => {
        test('works ok', () => {
            expect(transform([createConditions(null, null, null, null)], false)).toMatchSnapshot()
        })
    })

    describe('with conditions', () => {
        test('works ok', () => {
            expect(transform([createConditions(true, 'ios', {
                radius: 12,
                latLng: 'position'
            }, [
                {
                    id: 12,
                    condition: 'read',
                    value: false
                }
            ])], false)).toMatchSnapshot()
        })
    })
})
