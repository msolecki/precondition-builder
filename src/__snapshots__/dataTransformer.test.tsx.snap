// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dataTransformer empty conditions works ok 1`] = `
Object {
  "active": false,
  "conditions": Array [],
}
`;

exports[`dataTransformer with conditions works ok 1`] = `
Object {
  "active": false,
  "conditions": Array [
    Array [
      "client.tracking.system.os == 'ios'",
      "logged == 'true'",
      "location_radius == '12'",
      "public.kesi.additionalData.12.read == false",
    ],
  ],
}
`;
